firedrake/linear_solver.py

```diff
@@ -116,36 +118,27 @@ def _rhs(self):
 
         u = function.Function(self.trial_space)
         b = cofunction.Cofunction(self.test_space.dual())
-        expr = -action(self.A.a, u)
-        return u, get_assembler(expr).assemble, b
+        expr = - action(self.A.a, u)
+        return u, get_assembler(expr, bcs=self.A.bcs, zero_bc_nodes=False, needs_zeroing=False).assemble, b
 
     def _lifted(self, b):
         u, update, blift = self._rhs
         u.dat.zero()
         for bc in self.A.bcs:
             bc.apply(u)
+        blift.assign(b)
         update(tensor=blift)
-        # blift contains -A u_bc
-        blift += b
-        if isinstance(blift, cofunction.Cofunction):
-            blift_func = blift.riesz_representation(riesz_map="l2")
-            for bc in self.A.bcs:
-                bc.apply(blift_func)
-            blift.assign(blift_func.riesz_representation(riesz_map="l2"))
-        else:
-            for bc in self.A.bcs:
-                bc.apply(blift)
         # blift is now b - A u_bc, and satisfies the boundary conditions
         return blift
 
     @PETSc.Log.EventDecorator()
     def solve(self, x, b):
```

Comment:
Please can you add a docstring which spells out the type requirements on `x` and `b`.

Reply:
Done